src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.Factory.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.ComponentModel.Composition;
+using System.Reflection;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
+{
+    internal partial class VisualStudioDiagnosticAnalyzerProvider
+    {
+        /// <summary>
+        /// A separate factory class is required since <see cref="SVsServiceProvider"/> needs different version of MEF than <see cref="WorkspaceEventListener"/>.
+        /// </summary>
+        [Export(typeof(Factory))]
+        internal sealed class Factory
+        {
+            public VisualStudioDiagnosticAnalyzerProvider Provider { get; }
+
+            [ImportingConstructor]
+            [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+            public Factory([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+            {
+                var dte = (EnvDTE.DTE)serviceProvider.GetService(typeof(EnvDTE.DTE));
```

Comment:
So this is bad to do in a constructor since this can be constructed off the UI thread, and this would marshal back. VS MEF does not follow VS Threading rules so that absolutely could deadlock. You're best off just holding onto the IServiceProvider and moving this into a method, and don't call that until inside your fire-and-forget method.